File: app/src/main/java/com/hhp227/application/Tab4Fragment.java
package com.hhp227.application;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import com.android.volley.*;
import com.android.volley.toolbox.JsonObjectRequest;
import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.hhp227.application.activity.FeedbackActivity;
import com.hhp227.application.activity.MyinfoActivity;
import com.hhp227.application.activity.VerInfoActivity;
import com.hhp227.application.app.AppController;
import com.hhp227.application.app.URLs;
import com.hhp227.application.dto.User;
import com.hhp227.application.fragment.GroupFragment;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class Tab4Fragment extends Fragment implements View.OnClickListener {
    private static final String TAG = Tab4Fragment.class.getSimpleName();
    private boolean mIsAuth;
    private int mGroupId, mAuthorId;
    private String name, email, profile_img;

    public static Tab4Fragment newInstance(int groupId, int authorId) {
        Tab4Fragment fragment = new Tab4Fragment();
        Bundle args = new Bundle();

        args.putInt("group_id", groupId);
        args.putInt("author_id", authorId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mGroupId = getArguments().getInt("group_id");
            mAuthorId = getArguments().getInt("author_id");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_tab, container, false);
        SwipeRefreshLayout swipeRefreshLayout = rootView.findViewById(R.id.swipe_refresh_layout);
        RecyclerView recyclerView = rootView.findViewById(R.id.recycler_view);

        swipeRefreshLayout.setEnabled(false);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(new RecyclerView.Adapter<ViewHolder>() {
            @Override
            public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
                return new ViewHolder(LayoutInflater.from(getContext()).inflate(R.layout.fragment_tab4, parent, false));
            }

            @Override
            public void onBindViewHolder(ViewHolder holder, int position) {
                User user = AppController.Companion.getInstance().getPreferenceManager().getUser();
                mIsAuth = user.getId() == mAuthorId;

                name = user.getName();
                email = user.getEmail();
                profile_img = user.getProfileImage();

                holder.txtName.setText(name);
                holder.txtEmail.setText(email);
                Glide.with(getActivity())
                        .load(URLs.URL_USER_PROFILE_IMAGE + profile_img)
                        .apply(RequestOptions.errorOf(R.drawable.profile_img_circle).circleCrop())
                        .into(holder.profilePic);

                AdRequest adRequest = new AdRequest.Builder().build();
                holder.mAdView.loadAd(adRequest);
                holder.Profile.setOnClickListener(Tab4Fragment.this);
                holder.txtWithdrawal.setText("그룹" + (mIsAuth ? " 폐쇄" : " 탈퇴"));
                holder.WithDrawal.setOnClickListener(Tab4Fragment.this);
                holder.Appstore.setOnClickListener(Tab4Fragment.this);
                holder.Feedback.setOnClickListener(Tab4Fragment.this);
                holder.Version.setOnClickListener(Tab4Fragment.this);
                holder.Share.setOnClickListener(Tab4Fragment.this);
            }

            @Override
            public int getItemCount() {
                return 1;
            }
        });

        return rootView;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.profile:
                profile();
                break;
            case R.id.llWithdrawal:
                withdrawal();
                break;
            case R.id.appstore:
                appstore();
                break;
            case R.id.feedback:
                feedback();
                break;
            case R.id.share:
                share();
                break;
            case R.id.verinfo:
                verinfo();
                break;
        }
    }

    private void share() {
        Intent share = new Intent(Intent.ACTION_SEND);
        share.setType("text/plain");
        share.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        share.putExtra(Intent.EXTRA_SUBJECT,
                getString(R.string.app_name));
        share.putExtra(Intent.EXTRA_TEXT, "확인하세요" + "\n" +
                "GitHub Page :  https://localhost/" +
                "Sample App : https://play.google.com/store/apps/details?id=");
        startActivity(Intent.createChooser(share,
                getString(R.string.app_name)));
    }

    private void withdrawal() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());

        builder.setMessage((mIsAuth ? "폐쇄" : "탈퇴") + "하시겠습니까?");
        builder.setPositiveButton("예", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.DELETE, (mIsAuth ? URLs.URL_GROUP : URLs.URL_LEAVE_GROUP) + "/" + mGroupId, null, new Response.Listener<JSONObject>() {
                    @Override
                    public void onResponse(JSONObject response) {
                        try {
                            if (!response.getBoolean("error")) {
                                getActivity().setResult(Activity.RESULT_OK, new Intent(getContext(), GroupFragment.class));
                                getActivity().finish();
                                // 글쓰기나 글삭제후 그룹탈퇴하면 GroupFragment 목록이 새로고침이 되지 않음
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                }, new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        VolleyLog.e(TAG, error.getMessage());
                    }
                }) {
                    @Override
                    public Map<String, String> getHeaders() {
                        Map<String, String> headers = new HashMap<>();

                        headers.put("Authorization", AppController.Companion.getInstance().getPreferenceManager().getUser().getApiKey());
                        return headers;
                    }
                };

                AppController.Companion.getInstance().addToRequestQueue(jsonObjectRequest);
            }
        });
        builder.setNegativeButton("아니오", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
            }
        });
        builder.show();
    }

    private void appstore() {
        String appUrl = "https://play.google.com/store/apps/details?id=" + getContext().getPackageName();
        Intent rateIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(appUrl));
        startActivity(rateIntent);
    }

    private void feedback() {
        Intent intent = new Intent(getActivity(), FeedbackActivity.class);
        startActivity(intent);
    }

    private void profile() {
        Intent intent = new Intent(getActivity(), MyinfoActivity.class);
        startActivity(intent);
    }

    private void verinfo() {
        Intent intent = new Intent(getActivity(), VerInfoActivity.class);
        startActivity(intent);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private ImageView profilePic;
        private TextView txtName, txtEmail, txtWithdrawal;
        private LinearLayout Profile, WithDrawal, Version, Feedback, Appstore, Share;
        private AdView mAdView;

        ViewHolder(View itemView) {
            super(itemView);
            profilePic = itemView.findViewById(R.id.profilePic);
            txtName = itemView.findViewById(R.id.pname);
            txtEmail = itemView.findViewById(R.id.pemail);
            Profile = itemView.findViewById(R.id.profile);
            WithDrawal = itemView.findViewById(R.id.llWithdrawal);
            txtWithdrawal = itemView.findViewById(R.id.tvWithdrawal);
            Version = itemView.findViewById(R.id.verinfo);
            Appstore = itemView.findViewById(R.id.appstore);
            Feedback = itemView.findViewById(R.id.feedback);
            Share = itemView.findViewById(R.id.share);
            mAdView = itemView.findViewById(R.id.adView);
        }
    }
}

File: app/src/main/java/com/hhp227/application/fragment/Tab4Fragment.kt
package com.hhp227.application.fragment

import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.volley.Response
import com.android.volley.VolleyLog
import com.android.volley.toolbox.JsonObjectRequest
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.android.gms.ads.AdRequest
import com.hhp227.application.R
import com.hhp227.application.activity.FeedbackActivity
import com.hhp227.application.activity.MyInfoActivity
import com.hhp227.application.activity.VerInfoActivity
import com.hhp227.application.app.AppController
import com.hhp227.application.app.URLs
import com.hhp227.application.databinding.FragmentTab4Binding
import com.hhp227.application.databinding.FragmentTabBinding
import com.hhp227.application.dto.User
import com.hhp227.application.util.autoCleared
import org.json.JSONException

class Tab4Fragment : Fragment(), View.OnClickListener {
    private var groupId = 0

    private var authorId = 0

    private var isAuth = false

    private var binding: FragmentTabBinding by autoCleared()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            groupId = it.getInt(ARG_PARAM1)
            authorId = it.getInt(ARG_PARAM2)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentTabBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.swipeRefreshLayout.isRefreshing = false

        binding.recyclerView.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = object : RecyclerView.Adapter<ViewHolder>() {
                override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
                    return ViewHolder(FragmentTab4Binding.inflate(LayoutInflater.from(parent.context), parent, false))
                }

                override fun onBindViewHolder(holder: ViewHolder, position: Int) {
                    holder.bind(AppController.getInstance().preferenceManager.user)
                }

                override fun getItemCount(): Int = 1
            }
        }
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.profile -> Intent(requireContext(), MyInfoActivity::class.java).also(::startActivity)
            R.id.ll_withdrawal -> AlertDialog.Builder(requireContext()).setMessage(if (isAuth) "폐쇄" else "탈퇴" + "하시겠습니까?")
                .setPositiveButton("예") { _, _ ->
                    val jsonObjectRequest: JsonObjectRequest = object : JsonObjectRequest(
                        Method.DELETE,
                        "${if (isAuth) URLs.URL_GROUP else URLs.URL_LEAVE_GROUP}/$groupId",
                        null,
                        Response.Listener { response ->
                            try {
                                if (!response.getBoolean("error")) {
                                    requireActivity().setResult(Activity.RESULT_OK, Intent(context, GroupFragment::class.java))
                                    requireActivity().finish()
                                    // 글쓰기나 글삭제후 그룹탈퇴하면 GroupFragment 목록이 새로고침이 되지 않음
                                }
                            } catch (e: JSONException) {
                                e.printStackTrace()
                            }
                        },
                        Response.ErrorListener { error ->
                            VolleyLog.e(TAG, error.message)
                        }) {
                        override fun getHeaders() = mapOf("Authorization" to AppController.getInstance().preferenceManager.user.apiKey)
                    }

                    AppController.getInstance().addToRequestQueue(jsonObjectRequest)
                }
                .setNegativeButton("아니오") { dialog, _ -> dialog.dismiss() }
                .show()
            R.id.appstore -> Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=${requireContext().packageName}")).also(::startActivity)
            R.id.feedback -> Intent(requireContext(), FeedbackActivity::class.java).also(::startActivity)
            R.id.verinfo -> Intent(requireContext(), VerInfoActivity::class.java).also(::startActivity)
            R.id.share -> startActivity(Intent.createChooser(Intent().let { intent ->
                intent.setType("text/plain").addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.app_name))
                intent.putExtra(Intent.EXTRA_TEXT,
                    """
     확인하세요
     GitHub Page :  https://localhost/Sample App : https://play.google.com/store/apps/details?id=
     """.trimIndent()
                )
            }, getString(R.string.app_name)))
        }
    }

    inner class ViewHolder(private val binding: FragmentTab4Binding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(user: User) = with(binding) {
            isAuth = user.id == authorId
            pname.text = user.name
            pemail.text = user.email
            tvWithdrawal.text = "그룹" + if (isAuth) "폐쇄" else "탈퇴"

            Glide.with(binding.root)
                .load(URLs.URL_USER_PROFILE_IMAGE + user.profileImage)
                .apply(RequestOptions.errorOf(R.drawable.profile_img_circle).circleCrop())
                .into(ivProfileImage)
            adView.loadAd(AdRequest.Builder().build())
        }

        init {
            with(binding) {
                profile.setOnClickListener(::onClick)
                llWithdrawal.setOnClickListener(::onClick)
                appstore.setOnClickListener(::onClick)
                feedback.setOnClickListener(::onClick)
                verinfo.setOnClickListener(::onClick)
                share.setOnClickListener(::onClick)
            }
        }
    }

    companion object {
        private const val ARG_PARAM1 = "group_id"
        private const val ARG_PARAM2 = "author_id"
        private val TAG = Tab4Fragment::class.java.simpleName

        fun newInstance(groupId: Int, authorId: Int) = Tab4Fragment().apply {
            arguments = Bundle().apply {
                putInt(ARG_PARAM1, groupId)
                putInt(ARG_PARAM2, authorId)
            }
        }
    }
}